Use official phpfpm

fjmk opened the pull request:
I have made a start with refactoring service-cli to use the official docker php-fmp images. I think everything is available in version 5.6.

There are some small differences:

- DTrace Support => disabled  (apparently this has to compiled in the php binary?)
- the mhash extension is not available (drupal doesn't need mhash)
- some version differences in the extensions

When you think this is a good start, I would like to refactor this repo further:
Change the direcotory structure (like official docker repo):

- 5.6/
- 7.0/
- 7.1/
- 7.2-rc/

Let me know what you think.

lmakarov commented:
Oh, nice!

I actually started working on that same thing some time ago with the goal to have php7.1 support out in the next Docksal release. Good timing!

I'd take a look at post my feedback later next week.

fjmk commented:
I went on with the official PHP docker images.

- moved files to /5.6/ directory
- created 7.0, 7.1 and 7.2 directories with their own Dockerfile, scripts and config
  - I don't know if this is the best option, no difference in Dockerfile between 7.0 and 7.1
  - But it is easy for auto-update to dockerhub, i believe

Some important changes in PHP 7.x images:

- Removed **memcache** for PHP 7.x. See https://groups.drupal.org/node/513801 for reasoning.
- **mcrypt** is removed from PHP 7.2.x See https://wiki.php.net/rfc/mcrypt-viking-funeral (is depreciated in 7.1.x)
- **xdebug** temporary removed, pecl version not ready for PHP 7.2 yet? I _can_ try to compile xdebug manualy for 7.2.x.

NOTE: ghostscript has to be added!

cferthorney commented:
Is there anything I can do to help get this merged in? Or are we simply awaiting final acceptance and merge into master and images pushed to Docker Hub?

lmakarov commented:
Unfortunately, I have not been able to take a closer look at this PR yet. It changes the whole repo structure and also the underlying images. There are many things that can go wrong here.

@cferthorney if you are able to build and test the images, then provide feedback here, that would be much appreciated.

cferthorney commented:
@lmakarov of course, I will try to get that done over this weekend.  Do I need to build them within the VM and then just update my `docksal-local.env` file?

fjmk commented:
@cferthorney Nice you volunteer to test this pull request. 

I assume here you want to test php-fpm version 7.1..

Go to your project path and execute the commands:  
```
cd /path/of/your/project
git clone --branch use-official-phpfpm git@github.com:fjmk/service-cli.git
cd service-cli/7.1
fin docker build -t docksal:2.0-php71 .       # Take a nice cup of Coffee
cd ../..
```

Now you can change in your ```docksal-local.env``` file:  
```CLI_IMAGE='docksal:2.0-php71'```  
and run:  
``` fin up```

You can repeat this for the service-cli/5.6 and service-cli/7.0 versions.

Success

cferthorney commented:
I've tested the PHP 7.1 and PHP 5.6 on a [Drupal Composer project](https://github.com/drupal-composer/drupal-project) and a Wordpress site.  PHP runs, reports itself nicely.  My one concern is the image sizes have increased as follows:

- 1.2  GB for PHP 7.1
- 1.19 GB for PHP 7.0
-  1.0 GB for PHP 5.6

However, I don't feel it right I review what has been included in the PHP builds.  If someone who has a better understanding of what Docksal requires from a PHP perspective that may be a good idea.  As is, I believe this PR should work as advertised and could be merged in.

cferthorney commented:
@fjmk thanks - I will redownload and test again later today (07 Nov 2017).

achekulaev commented:
@lmakarov I think we should prioritize reviewing this and getting it merged. However we should merge it into develop first.

fjmk commented:
I have managed to make the images ~200Mb smaller.

Some things to note: 
1. Merge changes since Fri Sep 22 in develop branche
2. Change the [PHP override documentation](https://docksal.readthedocs.io/en/master/advanced/stack-settings/#php-and-mysql-configuration) and or the [example code](https://github.com/docksal/docksal/tree/develop/examples/.docksal/etc/php)

 See [Dockerfile line 176](https://github.com/fjmk/service-cli/blob/ab55f60762971c9b604f20d0e1c1bf17b1ab6b6f/5.6/Dockerfile#L176)

fjmk commented:
@lmakarov Merged develop and changed the destination branch.

I am busy with the image size, hope to commit this weekend.

fjmk commented:
I remove the php build dependencies now, as done in the upstream image. We save another ~40MB with this. 

With [docker-squash](https://github.com/jwilder/docker-squash) we **can** get smaller images (~86MB) but I don't feel like using this to shrink the images.

As a plus I created Makefiles & bats tests:
- one main Makefile to _build|test|clean_ all images with one command
- a Makefile for every version with the same goals as in your service-web repository

Needs review.

lmakarov commented:
@fjmk great! Going to review now.

> With docker-squash we can get smaller images (~86MB)
Is that an additional reduction in size by 86MB or the total image size of just 86MB (which sounds pretty unrealistic).

lmakarov commented on tests/test.bats:
```diff
@@ -0,0 +1,234 @@
+	export REQUEST_URI=/
+	export QUERY_STRING=
+	export REQUEST_METHOD=GET
+	run cgi-fcgi -bind -connect test.docksal:19000
```
`cgi-fcgi` is not available on macOS, so I cannot test it directly.

Thinking about other options that would works cross platform:
- test just PHP CLI, not FPM
- test using [PHP built-in web server](http://php.net/manual/en/features.commandline.webserver.php)

lmakarov commented:
I compared `php -i` and `php -m` output for the existing images and the new ones (php5.6 and php7.0). Diff summaries: https://gist.github.com/lmakarov/9513e0a5d2e6e72616fc16665120b8af

Some of those may introduce breaking changes and need to be adjusted appropriately. TBD which ones.

We should probably start bundling compete `php.ini` files with our images, based on the stock php.ini-development versions (+ our adjustments). This would make it much easier to review the resulting configuration vs the approach we've been using with `sed` at build time (in Dockerfile).

https://github.com/php/php-src/blob/PHP-5.6.32/php.ini-development
https://github.com/php/php-src/blob/PHP-7.0.25/php.ini-development

I really like the way Wodby (Docker4Drupal) images are organized, including settings templates overrides via environment variables, instead of files: https://github.com/wodby/php

Unfortunately, Wodby's PHP images are based on Alpine and this not very useful for local development - it's bare bones and not familiar at all for most.

lmakarov commented:
@fjmk Big thanks for adding tests! I set up TravisCI build configuration to run them. Unfortunately, they are failing for all image versions at the moment. Please review.

fjmk commented:
@lmakarov Thanks for the extensive review and sorry about faulty container names in the tests.

All TravisCI tests have passed now, very nice :)

The way Wodby is doing the configuration by environment variables is nice, but the docsal[-local].env could become very big. For me, Docksal's solution with overrides is more clear.

I am not sure where/how I can help now.

Maybe we can setup a one-to-one gitter chat? I am on CET time and available tonight between 20.00 and 22:00 CET. If yes, propose a time, please.

-- Frans

lmakarov commented:
@fjmk Unfortunately I won't have time today within the window you proposed. Here are some notes in the mean time:

- PHP 7.1 build keeps failing - https://travis-ci.org/docksal/service-cli/jobs/301277046
- We should include memcache extension back into the PHP7.x images. Looks like it has a stable [Pecl release](https://pecl.php.net/package/memcached) and Acquia does use memcache with PHP7 - https://docs.acquia.com/article/support-php-71-acquia-cloud
- Move all settings overrides from inline sed command in Dockerfile into a php.ini and php-fpm.conf (names can be different) files respectively, similar to how Wodby does it with [docker-php.ini.tpl](https://github.com/wodby/php/blob/master/7.0/templates/docker-php.ini.tpl) and [zz-www.conf.tpl](https://github.com/wodby/php/blob/master/7.0/templates/zz-www.conf.tpl). We don't have to use their templating approach for now, just flat files with values.

achekulaev commented:
@fjmk @lmakarov just wanted to let you guys know that I excitedly follow this conversation!

fjmk commented:
@achekulaev nice to hear :)

- PHP 7.1 build keeps failing - https://travis-ci.org/docksal/service-cli/jobs/301277046

I hope te have solved the random failing of travisCI builds bij adding ```pecl update-channels``` Hmmm.. apparently not.... I feel that rerunning the tests will succeed (most of the time).
 
- We should include the Pecl memcached extension into the PHP7.x images. Looks like it has a stable release and Acquia does use it with PHP7
https://docs.acquia.com/article/support-php-71-acquia-cloud
https://gist.github.com/lmakarov/50bb06008a1aabfed3ffa768a041853d#file-php7-1-modules-txt-L30

See commit 9a5b92e

- Move all settings overrides from inline sed commands in Dockerfile into a php.ini and php-fpm.conf (names can be different) files respectively, similar to how Wodby does it with docker-php.ini.tpl and zz-www.conf.tpl. We don't have to use their templating approach for now, just flat files with values.

Have researched a lot of possibilities and came up with the solution in commit  e3ae291
What have I done:
- We didn't have a php.ini file like we had in the debian version. A number of differences in https://gist.github.com/lmakarov/9513e0a5d2e6e72616fc16665120b8af where the reason for this.
I found https://raw.githubusercontent.com/php/php-src/PHP-${PHP_VERSION}/php.ini-development and use this as our default php.ini. There are some minor differences with the debian php.ini, only ```error_reporting = E_ALL``` (better for development) and ```mysqlnd.collect_statistics = On```. In the debian FPM php.ini the differences with the CLI are the ```expose_php = Off``` and ```disable_functions = pcntl_alarm,pcntl_fork,pcntl_waitpid,pcntl_wait,pcntl_wifexited,pcntl_wifstopped,pcntl_wifsignaled,pcntl_wexitstatus,pcntl_wtermsig,pcntl_wstopsig,pcntl_signal,pcntl_signal_dispatch,pcntl_get_last_error,pcntl_strerror,pcntl_sigprocmask,pcntl_sigwaitinfo,pcntl_sigtimedwait,pcntl_exec,pcntl_getpriority,pcntl_setpriority,```  settings
- In the default docksal-php.ini file are the settings we did in the sed commands. The echo/sed commands te change settings in ```opcache``` and ```blackfire``` are in this same file. This limits the number of settings files to maintain.
- In debian the CLI and FPM have their own php.ini. This is not available in the official images. Ini settings in ```php.ini``` and conf.d/[name].ini affects FPM too. The FPM settings are done in docksal-www.conf file.
- User  overrides:
.docksal/etc/php/php-cli.ini settings for CLI **and** FPM (unless overridden in docksal-www.conf or php-fpm.conf)
.docksal/etc/php/php-fpm.conf settings for FPM only.
- I have added the mysql and mysqli extensions in the 5.6 image.

Let me know what you think. The changes for the 7.x

fjmk commented:
No comments on my last comment, so I went ahead and made the same changes for 7.0, 7.1 and 7.2.

Travis tests are not too reliable, so please request a re-test of the failing test.

lmakarov commented:
@fjmk thanks. I did some refactoring over the weekend, but did not get to pushing it up. I'll try to get that out of the door this week and merged into develop.

As for Travis builds, either http://pecl.php.net is doing some throttling or it is not a very reliable source, as these happen randomly:

```
Could not download from "http://pecl.php.net/get/gnupg-1.4.0.tgz", cannot download "pecl/gnupg" (Connection to `pecl.php.net:80' failed: Connection timed out)
```

Let's do what [Wodby does](https://github.com/wodby/php/blob/master/.travis.yml#L34) with [travis_retry](https://docs.travis-ci.com/user/common-build-problems/#travis_retry):

```
- travis_retry make && make test
```

lmakarov commented:
The original objective of the PR is now achieved - time to merge!
@fjmk many thanks for your contributions!

drubage commented:
Is this officially merged yet? Will the default php cli be 7.1 now or how do we specify that we want 7.1?

lmakarov commented:
Edge version of the new 2.0 images are here! 

```
$ fin image registry docksal/cli | grep 'edge.*\.'
docksal/cli:edge-php7.2
docksal/cli:edge-php5.6
docksal/cli:edge-php7.1
docksal/cli:edge-php7.0
```

Feel free to start using them and report any issues in this repo. We will be releasing them and including in the standard stacks with the next Docksal minor release.

lmakarov commented:
@drubage 

> Will the default php cli be 7.1 now or how do we specify that we want 7.1?

PHP7.1 will be used in the default and acquia stacks

drubage commented:
Radical!
